src/modules/function/declaration.rs

```diff
@@ -145,14 +145,30 @@ impl SyntaxModule<ParserMetadata> for FunctionDeclaration {
                     Err(_) => token(meta, ",")?
                 };
             }
+
+            let mut type_tok = None;
+            let mut marked_as_failable: Option<bool> = None;
             // Optionally parse the return type
             match token(meta, ":") {
-                Ok(_) => self.returns = parse_type(meta)?,
+                Ok(_) => {
+                    type_tok = Some(meta.get_current_token());
+                    self.returns = parse_type(meta)?;
+                    marked_as_failable = Some(token(meta, "?").is_ok());
+                },
                 Err(_) => self.returns = Type::Generic
             }
+
+
             // Parse the body
             token(meta, "{")?;
-            let (index_begin, index_end, is_failable) = skip_function_body(meta);
+            let (index_begin, index_end, mut is_failable) = skip_function_body(meta);
+            if let (Some(marked_as_failable), Some(type_tok)) = (marked_as_failable, type_tok) {
```

Comment:
I think that it’s more readable when it’s:

```
if function_is_failable and !marked_as_failable:
    error()
```


Something like:

```suggestion
            if let Some(type_tok) = type_tok {
                if !marked_as_failable {
```

And then the appropriate indentation and closed bracket.